<a href="https://colab.research.google.com/github/optimalMachine/EliteDataScienceCourse/blob/master/5_Workbook_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### COMPANION WORKBOOK

# Pandas

To make the most out this program, we strongly recommend you to:
1. First practice writing and implementing all of the code from Coding Section of the online lesson.
2. Then, freely experiment with and explore any interesting or confusing concepts. Simply insert new code cells and then use the help of Google and official documentation.
3. Finally, tackle all of the exercises at the end. They will help you tie everything together and **learn in context.**

#### <span style="color:#555">LESSON CODE SANDBOX</span>

Use this space to practice writing and implementing all of the code from Coding Section of the online lesson. Insert new code cells as needed, and feel free to write notes to yourself in Markdown.

## I. Pandas DataFrames are like spreadsheets.

In [50]:
import pandas as pd

In [51]:
example_dataframe = pd.DataFrame({'column_1':[5,4,3],'column_2':['a','b','c']})
example_dataframe

column_1 column_2
0         5        a
1         4        b
2         3        c

In [52]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [53]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Python for Data Science/project_files/iris.csv')
print (type(df))

<class 'pandas.core.frame.DataFrame'>


In [54]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [55]:
print (df.shape)

(150, 5)


In [56]:
print (len(df))
print (df.shape[0])

150
150


In [57]:
print (df.min())

sepal_length       4.3
sepal_width        2.0
petal_length       1.0
petal_width        0.1
species         setosa
dtype: object


In [58]:
print (df.max())

sepal_length          7.9
sepal_width           4.4
petal_length          6.9
petal_width           2.5
species         virginica
dtype: object


In [59]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

## II. Pandas Series are like single columns.

In [60]:
integer_series = pd.Series([0,1,2,3,4])
print (integer_series)
print (type(integer_series))

0    0
1    1
2    2
3    3
4    4
dtype: int64
<class 'pandas.core.series.Series'>


In [61]:
print (type(df.petal_length))

<class 'pandas.core.series.Series'>


In [62]:
print (type(df['petal_length']))

<class 'pandas.core.series.Series'>


In [63]:
print (df.petal_length.head())
print ('The minimum petal length is', df.petal_length.min())
print ('The maximum petal length is', df.petal_length.max())

0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: petal_length, dtype: float64
The minimum petal length is 1.0
The maximum petal length is 6.9


In [64]:
print (df.species.unique())

['setosa' 'versicolor' 'virginica']


In [65]:
print (df.petal_length.median())
import numpy as np
print (np.median(df.petal_length))

4.35
4.35


In [66]:
df['petal_area'] = df.petal_width * df.petal_length
df['petal_area'].head()

0    0.28
1    0.28
2    0.26
3    0.30
4    0.28
Name: petal_area, dtype: float64

## III. Boolean masks allow you to filter data.

In [67]:
list_mask = [True,True,True,True]
series_mask = ([True,True,False,False])

In [68]:
print (series_mask)

[True, True, False, False]


In [69]:
example_series = pd.Series([10,5,-3,2])
print (example_series)

0    10
1     5
2    -3
3     2
dtype: int64


In [70]:
series_mask = example_series > 0
print (series_mask)

0     True
1     True
2    False
3     True
dtype: bool


In [71]:
example_series[series_mask]

0    10
1     5
3     2
dtype: int64

In [72]:
example_series[~series_mask]

2   -3
dtype: int64

In [73]:
df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species',
       'petal_area'],
      dtype='object')

In [74]:
df[df.petal_area>14]

sepal_length  sepal_width  petal_length  petal_width    species  \
100           6.3          3.3           6.0          2.5  virginica   
109           7.2          3.6           6.1          2.5  virginica   
117           7.7          3.8           6.7          2.2  virginica   
118           7.7          2.6           6.9          2.3  virginica   
135           7.7          3.0           6.1          2.3  virginica   
144           6.7          3.3           5.7          2.5  virginica   

     petal_area  
100       15.00  
109       15.25  
117       14.74  
118       15.87  
135       14.03  
144       14.25

In [75]:
example_mask = pd.Series([True,False,False,True,False])
print (example_mask.astype(int))

0    1
1    0
2    0
3    1
4    0
dtype: int64


In [76]:
df.columns
df['giant'] = (df.petal_area>14).astype(int)
df[['petal_area','giant']].head()

petal_area  giant
0        0.28      0
1        0.28      0
2        0.26      0
3        0.30      0
4        0.28      0

In [77]:
species_mask = df.species.isin(['versicolor','virginica'])
sepal_mask = df.sepal_width > 3.2
df[species_mask & sepal_mask]


sepal_length  sepal_width  petal_length  petal_width     species  \
56            6.3          3.3           4.7          1.6  versicolor   
85            6.0          3.4           4.5          1.6  versicolor   
100           6.3          3.3           6.0          2.5   virginica   
109           7.2          3.6           6.1          2.5   virginica   
117           7.7          3.8           6.7          2.2   virginica   
124           6.7          3.3           5.7          2.1   virginica   
131           7.9          3.8           6.4          2.0   virginica   
136           6.3          3.4           5.6          2.4   virginica   
144           6.7          3.3           5.7          2.5   virginica   
148           6.2          3.4           5.4          2.3   virginica   

     petal_area  giant  
56         7.52      0  
85         7.20      0  
100       15.00      1  
109       15.25      1  
117       14.74      1  
124       11.97      0  
131       12.80      0  
136       13.44      0  
144       14.25      1  
148       12.42      0

## IV. Groupbys allow you to segment and aggregate data.

In [78]:
df.groupby('species').mean()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa             5.006        3.428         1.462        0.246      0.3656   
versicolor         5.936        2.770         4.260        1.326      5.7204   
virginica          6.588        2.974         5.552        2.026     11.2962   

            giant  
species            
setosa       0.00  
versicolor   0.00  
virginica    0.12

In [79]:
df.groupby('species').median()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa               5.0          3.4          1.50          0.2       0.300   
versicolor           5.9          2.8          4.35          1.3       5.615   
virginica            6.5          3.0          5.55          2.0      11.445   

            giant  
species            
setosa        0.0  
versicolor    0.0  
virginica     0.0

In [80]:
df.groupby('species').max()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa               5.8          4.4           1.9          0.6        0.96   
versicolor           7.0          3.4           5.1          1.8        8.64   
virginica            7.9          3.8           6.9          2.5       15.87   

            giant  
species            
setosa          0  
versicolor      0  
virginica       1

In [81]:
df.groupby('species').std()

sepal_length  sepal_width  petal_length  petal_width  petal_area  \
species                                                                        
setosa          0.352490     0.379064      0.173664     0.105386    0.181155   
versicolor      0.516171     0.313798      0.469911     0.197753    1.368403   
virginica       0.635880     0.322497      0.551895     0.274650    2.157412   

               giant  
species               
setosa      0.000000  
versicolor  0.000000  
virginica   0.328261

In [82]:
df.groupby('species').agg(['min','median','max'])

sepal_length             sepal_width             petal_length  \
                    min median  max         min median  max          min   
species                                                                    
setosa              4.3    5.0  5.8         2.3    3.4  4.4          1.0   
versicolor          4.9    5.9  7.0         2.0    2.8  3.4          3.0   
virginica           4.9    6.5  7.9         2.2    3.0  3.8          4.5   

                       petal_width             petal_area                 \
           median  max         min median  max        min  median    max   
species                                                                    
setosa       1.50  1.9         0.1    0.2  0.6       0.11   0.300   0.96   
versicolor   4.35  5.1         1.0    1.3  1.8       3.30   5.615   8.64   
virginica    5.55  6.9         1.4    2.0  2.5       7.50  11.445  15.87   

           giant             
             min median max  
species                      
setosa         0    0.0   0  
versicolor     0    0.0   0  
virginica      0    0.0   1

#### <span style="color:#555">EXERCISES</span>

Complete each of the following exercises.

First, remember to import Pandas and read in the iris dataset.

In [83]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [84]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Python for Data Science/project_files/iris.csv')

## <span style="color:RoyalBlue">Exercise 5.1 - Flowers for Casey</span>

In the previous lesson, we discovered that the first local business we need to visit in Park Royal is a local **flower shop.** When we arrive, we find out that the owner, Casey, urgently needs our help!

They just received a new shipment of **iris flowers**, but they've never stocked these flowers before. Casey asks us to share what we know about these new flowers. Luckily, we have the *iris dataset* to learn more about them ourselves!

In the previous lesson, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts. While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

#### A.) First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [85]:
# Create toy_df
toy_df = pd.concat([df.head(),df.tail()])


#### B.) Next, display <code style="color:steelblue">toy_df</code>.
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [86]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

<strong style="color:RoyalBlue">Expected output:</strong>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
    </tr>
  </tbody>
</table>

#### C.) Next, display a summary table for <code style="color:steelblue">toy_df</code>.
* It should have the mean, standard deviation, and quartiles for each of the columns.
* Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

In [87]:
# Describe toy_df
toy_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count     10.000000    10.000000     10.000000    10.000000
mean       5.590000     3.130000      3.290000     1.130000
std        0.807534     0.316403      1.995244     0.992248
min        4.600000     2.500000      1.300000     0.200000
25%        4.925000     3.000000      1.400000     0.200000
50%        5.500000     3.050000      3.250000     1.000000
75%        6.275000     3.350000      5.175000     1.975000
max        6.700000     3.600000      5.400000     2.300000

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
      <td>10.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>5.590000</td>
      <td>3.130000</td>
      <td>3.290000</td>
      <td>1.130000</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.807534</td>
      <td>0.316403</td>
      <td>1.995244</td>
      <td>0.992248</td>
    </tr>
    <tr>
      <th>min</th>
      <td>4.600000</td>
      <td>2.500000</td>
      <td>1.300000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>4.925000</td>
      <td>3.000000</td>
      <td>1.400000</td>
      <td>0.200000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>5.500000</td>
      <td>3.050000</td>
      <td>3.250000</td>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>6.275000</td>
      <td>3.350000</td>
      <td>5.175000</td>
      <td>1.975000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>6.700000</td>
      <td>3.600000</td>
      <td>5.400000</td>
      <td>2.300000</td>
    </tr>
  </tbody>
</table>

## <span style="color:RoyalBlue">Exercise 5.2 - Basic Feature Engineering</span>

Elementwise operations are very useful in machine learning, especially for **feature engineering.** Feature engineering is the process of creating new features (model input variables) from existing ones, and its one of the best ways for data scientists to add value.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

#### A.) First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [88]:
# Display petal_width and petal_length
df[['petal_width','petal_length']]

petal_width  petal_length
0            0.2           1.4
1            0.2           1.4
2            0.2           1.3
3            0.2           1.5
4            0.2           1.4
..           ...           ...
145          2.3           5.2
146          1.9           5.0
147          2.0           5.2
148          2.3           5.4
149          1.8           5.1

[150 rows x 2 columns]

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>petal_width</th>
      <th>petal_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.2</td>
      <td>1.3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.2</td>
      <td>1.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.2</td>
      <td>1.4</td>
    </tr>
    <tr>
      <th>145</th>
      <td>2.3</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>146</th>
      <td>1.9</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>147</th>
      <td>2.0</td>
      <td>5.2</td>
    </tr>
    <tr>
      <th>148</th>
      <td>2.3</td>
      <td>5.4</td>
    </tr>
    <tr>
      <th>149</th>
      <td>1.8</td>
      <td>5.1</td>
    </tr>
  </tbody>
</table>

#### B.) Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [98]:
# Create a new petal_area column
toy_df['petal_area'] = toy_df['petal_width']*toy_df['petal_length']

# Display toy_df
toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
  </tbody>
</table>

### C.) Finally, what do we now know about Iris flowers?

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do! Often, by creating new features, you can learn more about the data (and improve your machine learning models). As you might guess, a machine learning algorithm and take these features and tease out the underlying patterns in a much more precise and comprehensive way.

## <span style="color:RoyalBlue">Exercise 5.3 - Mask On Mask Off</span>

Now, let's use the <code style="color:steelblue">toy_df</code> to really drive home the concept of **boolean masks**, because we'll be using these very frequently in data science.

For example, assume we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

#### A.) First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.

In [101]:
# Display toy_df
toy_df

sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

#### B.) Take a look at the DataFrame and manually count the number that satisfy our conditions.
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

*In the toy_df, there are:*
* **3** observations with <code style="color:steelblue">petal_area > 10</code> (indices 145, 147, 148)
* **5** observations with <code style="color:steelblue">sepal_width > 3</code> (indices 0,2,3,4,148)
* **1** observation that satifies both conditions (index 148)

Great. Now we'll see what's going on under the hood when we use our boolean masks.

#### C.) Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [104]:
# Mask for petal_area > 10
petal_area_mask = toy_df.petal_area >10

# Display petal_area_mask
petal_area_mask

0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool
</pre>

#### D.) Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [105]:
# Mask for sepal_width > 3
sepal_width_mask = toy_df.sepal_width > 3

# Display sepal_width_mask
sepal_width_mask

0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool
</pre>

#### E.) Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.
* Note how their combination results in another boolean mask!

In [106]:
# Display both masks, combined
print (petal_area_mask & sepal_width_mask)

0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool


<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool
</pre>

#### F.) Finally, select the observation(s) from <code style="color:steelblue">toy_df</code> where both conditions are met.
* Do you get the result you expected based on (B)?

In [107]:
# Index with both masks
toy_df[petal_area_mask & sepal_width_mask]

sepal_length  sepal_width  petal_length  petal_width    species  \
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
148       12.42

## <span style="color:RoyalBlue">Exercise 5.4 - Iris Insights</span>

With the help of **groupbys**, we're almost ready to return to Casey and share what we've learned about Iris flowers. But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.


#### A.) First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [109]:
# Display all 'virginica' species, sorted by petal_area
toy_df[toy_df.species=='virginica'].sort_values(by='petal_area')

sepal_length  sepal_width  petal_length  petal_width    species  \
149           5.9          3.0           5.1          1.8  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
145           6.7          3.0           5.2          2.3  virginica   
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
149        9.18  
146        9.50  
147       10.40  
145       11.96  
148       12.42

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>149</th>
      <td>5.9</td>
      <td>3.0</td>
      <td>5.1</td>
      <td>1.8</td>
      <td>virginica</td>
      <td>9.18</td>
    </tr>
    <tr>
      <th>146</th>
      <td>6.3</td>
      <td>2.5</td>
      <td>5.0</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>9.50</td>
    </tr>
    <tr>
      <th>147</th>
      <td>6.5</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>10.40</td>
    </tr>
    <tr>
      <th>145</th>
      <td>6.7</td>
      <td>3.0</td>
      <td>5.2</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>11.96</td>
    </tr>
    <tr>
      <th>148</th>
      <td>6.2</td>
      <td>3.4</td>
      <td>5.4</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>12.42</td>
    </tr>
  </tbody>
</table>

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

#### B.) Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* Based on the output, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

In [112]:
# Display all 'setosa' species
toy_df[toy_df.species == 'setosa'].sort_values(by='petal_area')

sepal_length  sepal_width  petal_length  petal_width species  petal_area
2           4.7          3.2           1.3          0.2  setosa        0.26
0           5.1          3.5           1.4          0.2  setosa        0.28
1           4.9          3.0           1.4          0.2  setosa        0.28
4           5.0          3.6           1.4          0.2  setosa        0.28
3           4.6          3.1           1.5          0.2  setosa        0.30

<strong style="color:RoyalBlue">Expected output:</strong>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>petal_area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.26</td>
    </tr>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.28</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0.30</td>
    </tr>
  </tbody>
</table>

#### C.) Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.
* Do you get the same result?

In [113]:
# Median petal_area in toy_df
toy_df.groupby('species').petal_area.median()

species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64

<strong style="color:RoyalBlue">Expected output:</strong>
<pre>
species
setosa        0.28
virginica    10.40
Name: petal_area, dtype: float64
</pre>

#### Congratulations... You've completed the training mission!

In this lesson, you explored the Iris dataset using Pandas. After arming yourself with Pandas, you've successfully completed the rest of the training mission!*